### DISCUSSION TITLE
`once` accessor for lazy property initialization

### DISCUSSION THREAD

**Discussion by kasperk81** [dotnet/csharplang#6589] [Language Ideas]
flavors of this pattern are often found in csharp code:

```c#
private volatile bool _isSet;
private int _cache;

public int V
{
    get
    {
       if (!_isSet)
       {
            _cache = ExpensiveCall();
            _isSet = true;
       }
       return _cache;
    }
}
```

which initializes the value on first access. some use nullable backing field with coalescing operator `??`, others (who don't mind short-lived boxing) use `Lazy<T>` for same purpose.

### `once` accessor

can save users from reinventing the boilerplate:

```c#
public static int V
{
    get;
    once => ExpensiveCall();
    // V_once() will run before the first call to V_get(), and store the result in backing field
}
```

**Comment by HaloFour** [comment]
See: #312

**Comment by iam3yal** [comment]
Or you can just do something like this:
```csharp
private int? _cache;

public int V
    => _cache ??= ExpensiveCall();
```

**Reply by HaloFour** [reply]
There are a bunch of different ways to do this and each has different execution, publishing and synchronization behaviors.  For example, in this case `ExpensiveCall()` could be executed a bunch of times if multiple threads attempt to access that property at the same time.  This is why `Lazy<T>` has multiple publishing policies that you can affect via the constructor that accepts a [`LazyThreadSafetyMode`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.lazythreadsafetymode?view=net-6.0).

**Reply by iam3yal** [reply]
@HaloFour I might have misunderstood the proposal, is the OP suggesting that the language will have a shorthand for `Lazy` or something?

**Reply by HaloFour** [reply]
This is also why it's challenging to consider for a language feature, there is a lot of nuance to which publishing policy is appropriate and it's not "correct" for the language to just pick one.

See: https://github.com/dotnet/csharplang/issues/681#issuecomment-308587180

**Reply by kasperk81** [reply]
> it's not "correct" for the language to just pick one.

so `class Volatile` supports bunch of features, but `volatile` language support settled on one commonly used behavior.

if `once` accessor is called exactly once in a thread-safe manner, that would be a reasonable choice.

**Reply by jnm2** [reply]
@kasperk81 There is an existing conversation about the publishing policy, and a thumbs-down reaction on Halo's comment letting you know it exists isn't helping your goal.

**Reply by HaloFour** [reply]
The `volatile` language support is based on explicit runtime support through IL prefix opcodes, not through any BCL APIs.  Those helper methods were added much later.

Either way, it's the question of policy, and specifically not baking one specific policy into the language, that has dissuaded members of the language team from considering this.  That can't be hand-waved away.

**Reply by jnm2** [reply]
> if `once` accessor is called exactly once in a thread-safe manner, that would be a reasonable choice.

This is not what your code sample does at the top though; you need a lock to achieve this, such as what Lazy<T> does in that publishing mode. A lock is overhead that many people won't want.

**Reply by kasperk81** [reply]
> This is not what your code sample does at the top though

never said it was thread-safe

**Reply by HaloFour** [reply]
In any case there are existing discussions on this subject and the conversation should remain there.  I am giving you context as to why the team has shied away from adding this as a language feature.  You may disagree with it, but short of bringing something new to the conversation it's unlikely to convince anyone that those concerns don't matter.
